File: 0062-unique-paths/0062-unique-paths.java
class Solution {
    private int ways(int i,int j,int dp[][]){
        if(i ==0 && j ==0) return 1;
        if(i<0 || j < 0) return 0;
        if(dp[i][j] != -1) return dp[i][j]; 
        int up = ways(i-1,j,dp);
        int left = ways(i,j-1,dp);
        return dp[i][j] = up + left;
    }
    public int uniquePaths(int m, int n) {
        int dp[][] = new int[m][n];
        for(int row[]:dp){
            Arrays.fill(row,-1);
        }
       return ways(m-1,n-1,dp);
    }
}

File: 0062-unique-paths/0062-unique-paths.cpp
class Solution {
private:
    int f(int n,int m, vector<vector<int>> &dp) {
        if(n == 0 && m == 0) {
            return 1;
        }
        if(n < 0 || m < 0) {
            return 0;
        }
        if(dp[n][m] != -1) return dp[n][m];
        int up = f(n-1,m,dp);
        int left = f(n,m-1,dp);
        return dp[n][m] = up + left;
    }
public:
    int uniquePaths(int m, int n) {
        // vector<vector<int>> dp(n,vector<int>(m,0));
        vector<int> prev(m,0);
        vector<int> curr(m,0);
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if( i == 0 && j == 0) {
                    curr[j] = 1;
                } else {
                    int up = 0;
                    if(i > 0) {
                        up = prev[j];    
                    }
                    int left = 0;
                    if(j > 0) {
                        left = curr[j-1];
                    }
                    curr[j] = up + left;    
                }
            }
            prev = curr;
        }
        return prev[m-1];
    }
};